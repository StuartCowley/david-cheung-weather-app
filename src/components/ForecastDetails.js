import React from "react";
import PropTypes from "prop-types";
import "../styles/forecast-details.css";

function dateOrdinal(d) {
  let subfix;
  if (d === "31" || d === "21" || d === "01") {
    subfix = "st";
  } else if (d === "22" || d === "02") {
    subfix = "nd";
  } else if (d === "23" || d === "03") {
    subfix = "rd";
  } else {
    subfix = "th";
  }
  if (d.charAt(0) === "0") {
    return d.charAt(1) + subfix;
  }
  return d + subfix;
}

function ForecastDetails({ forecast }) {
  const { date, temperature, wind, humidity } = forecast;
  const tempDate1 = new Date(date).toDateString();
  const tempDate2 = tempDate1.split(" ");
  const formattedDate = `${tempDate2[0]} ${dateOrdinal(tempDate2[2])} ${
    tempDate2[1]
  }`;
  return (
    <div className="forecast-details" data-testid="forecast-details">
      <div className="forecast-details__date">{formattedDate}</div>

      <div className="forecast-details__temperature">
        Temperature: {temperature.min}&deg;C - {temperature.max}&deg;C
      </div>

      <div className="forecast-details__humidity">Humidity: {humidity}</div>
      <div className="forecast-details__wind">
        Wind: {wind.speed}mph {wind.direction.toUpperCase()}
      </div>
    </div>
  );
}

export default ForecastDetails;

ForecastDetails.propTypes = {
  forecast: PropTypes.shape({
    date: PropTypes.number,
    temperature: PropTypes.shape({
      max: PropTypes.number,
      min: PropTypes.number,
    }),
    wind: PropTypes.shape({
      speed: PropTypes.number,
      direction: PropTypes.string,
    }),
    humidity: PropTypes.number,
  }).isRequired,
};
