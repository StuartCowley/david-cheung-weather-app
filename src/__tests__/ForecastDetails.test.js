import React from "react";
import { render, screen } from "@testing-library/react";
import ForecastDetails from "../components/ForecastDetails";
import renderer from "react-test-renderer";

describe("ForecastDetails", () => {
  const validProps = {
    date: 1525046400000,
    temperature: {
      min: 12,
      max: 22,
    },
    humidity: 50,
    wind: {
      speed: 8,
      direction: "nne",
    },
  };

  test("Renders as expected", () => {
    const rendered = renderer.create(<ForecastDetails forecast={validProps} />);

    expect(rendered).toMatchSnapshot();
  });

  test("Assert all classname elements are present", () => {
    render(<ForecastDetails forecast={validProps} />);
    expect(document.querySelector(".forecast-details")).toBeTruthy();
    expect(document.querySelector(".forecast-details__date")).toBeTruthy();
    expect(
      document.querySelector(".forecast-details__temperature")
    ).toBeTruthy();
    expect(document.querySelector(".forecast-details__humidity")).toBeTruthy();
    expect(document.querySelector(".forecast-details__wind")).toBeTruthy();
  });

  test("Assert all props elements are present", () => {
    render(<ForecastDetails forecast={validProps} />);

    // We need at least 2 tests. One for snapshot testing, and one for values
    expect(screen.getByText("Mon 30th Apr")).toBeTruthy();
    expect(screen.getByText("Temperature: 12째C - 22째C")).toBeTruthy();
    expect(screen.getByText("Humidity: 50")).toBeTruthy();
    expect(screen.getByText("Wind: 8mph NNE")).toBeTruthy();

    expect(screen.getByText("Mon 30th Apr")).toBeInstanceOf(HTMLDivElement);
    expect(screen.getByText("Temperature: 12째C - 22째C")).toBeInstanceOf(
      HTMLDivElement
    );
    expect(screen.getByText("Humidity: 50")).toBeInstanceOf(HTMLDivElement);
    expect(screen.getByText("Wind: 8mph NNE")).toBeInstanceOf(HTMLDivElement);
  });
});
