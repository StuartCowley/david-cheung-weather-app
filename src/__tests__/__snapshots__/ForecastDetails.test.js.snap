// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDetails Renders as expected 1`] = `
<div
  className="forecast-details"
  data-testid="forecast-details"
>
  <div
    className="forecast-details__date"
  >
    Mon 30th Apr
  </div>
  <div
    className="forecast-details__temperature"
  >
    Temperature: 
    12
    °C - 
    22
    °C
  </div>
  <div
    className="forecast-details__humidity"
  >
    Humidity: 
    50
  </div>
  <div
    className="forecast-details__wind"
  >
    Wind: 
    8
    mph 
    NNE
  </div>
</div>
`;
